FaucetNailedFailoverForwardingTest test fails when run in combination with other tests

Seems we broke something with how output groups are used in #3476 

If you run the ``FaucetNailedFailoverForwardingOrderedTest`` and ``FaucetNailedFailoverForwardingTestcat`` back to back the second test will fail due to a group already exists error message:

```
port order: -o 2,3,0,1
Using config from /etc/faucet/hw_switch_config.yaml
Testing hardware, forcing test serialization
Logging test results in /var/tmp/faucet-tests-sakye6bn
test ports server started
adding test FaucetNailedFailoverForwardingOrderedTest
adding test FaucetNailedFailoverForwardingTest

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
running 0 tests in parallel and 2 tests serial
test_untagged (mininet_tests.FaucetNailedFailoverForwardingOrderedTest) ... *** Error setting resource limits. Mininet's performance may be affected.
ok
test_untagged (mininet_tests.FaucetNailedFailoverForwardingTest) ... FAIL

======================================================================
FAIL: test_untagged (mininet_tests.FaucetNailedFailoverForwardingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/faucet-src/clib/mininet_test_base.py", line 481, in tearDown
    self.assertFalse(oferrors, msg=oferrors)
AssertionError: "Mar 17 22:12:31 faucet.valve ERROR    DPID 222613276036429 (0xca77313d8d4d) faucet-1 OFError type: OFPET_GROUP_MOD_FAILED code: OFPGMFC_GROUP_EXISTS version=0x4,msg_type=0x1,msg_len=0x5c,xid=0xc369eeb2,OFPErrorMsg(code=0,data=bytearray(b'\\x04\\x0f\\x00P\\xc3i\\xee\\xb2\\x00\\x00\\x03\\x00\\x00\\x00\\x03\\xe9\\x00 \\x00\\x00\\x00\\x00\\x00\\x05\\xff\\xff\\xff\\xff\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\x00\\x05\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00 \\x00\\x00\\x00\\x00\\x00\\x02\\xff\\xff\\xff\\xff\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\x00\\x02\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00'),type=6) caused by version=0x4,msg_type=0xf,msg_len=0x50,xid=0xc369eeb2,OFPGroupMod(buckets=[OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=5,type=0)],len=32,watch_group=4294967295,watch_port=5,weight=0), OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=2,type=0)],len=32,watch_group=4294967295,watch_port=2,weight=0)],command=0,group_id=1001,type=3)\n\n\nMar 17 22:12:31 faucet.valve ERROR    DPID 222613276036429 (0xca77313d8d4d) faucet-1 OFError type: OFPET_GROUP_MOD_FAILED code: OFPGMFC_GROUP_EXISTS version=0x4,msg_type=0x1,msg_len=0x5c,xid=0xc369eeb3,OFPErrorMsg(code=0,data=bytearray(b'\\x04\\x0f\\x00P\\xc3i\\xee\\xb3\\x00\\x00\\x03\\x00\\x00\\x00\\x03\\xea\\x00 \\x00\\x00\\x00\\x00\\x00\\x05\\xff\\xff\\xff\\xff\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\x00\\x05\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00 \\x00\\x00\\x00\\x00\\x00\\x02\\xff\\xff\\xff\\xff\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x10\\x00\\x00\\x00\\x02\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00'),type=6) caused by version=0x4,msg_type=0xf,msg_len=0x50,xid=0xc369eeb3,OFPGroupMod(buckets=[OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=5,type=0)],len=32,watch_group=4294967295,watch_port=5,weight=0), OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=2,type=0)],len=32,watch_group=4294967295,watch_port=2,weight=0)],command=0,group_id=1002,type=3)\n" is not false : Mar 17 22:12:31 faucet.valve ERROR    DPID 222613276036429 (0xca77313d8d4d) faucet-1 OFError type: OFPET_GROUP_MOD_FAILED code: OFPGMFC_GROUP_EXISTS version=0x4,msg_type=0x1,msg_len=0x5c,xid=0xc369eeb2,OFPErrorMsg(code=0,data=bytearray(b'\x04\x0f\x00P\xc3i\xee\xb2\x00\x00\x03\x00\x00\x00\x03\xe9\x00 \x00\x00\x00\x00\x00\x05\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x02\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00'),type=6) caused by version=0x4,msg_type=0xf,msg_len=0x50,xid=0xc369eeb2,OFPGroupMod(buckets=[OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=5,type=0)],len=32,watch_group=4294967295,watch_port=5,weight=0), OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=2,type=0)],len=32,watch_group=4294967295,watch_port=2,weight=0)],command=0,group_id=1001,type=3)


Mar 17 22:12:31 faucet.valve ERROR    DPID 222613276036429 (0xca77313d8d4d) faucet-1 OFError type: OFPET_GROUP_MOD_FAILED code: OFPGMFC_GROUP_EXISTS version=0x4,msg_type=0x1,msg_len=0x5c,xid=0xc369eeb3,OFPErrorMsg(code=0,data=bytearray(b'\x04\x0f\x00P\xc3i\xee\xb3\x00\x00\x03\x00\x00\x00\x03\xea\x00 \x00\x00\x00\x00\x00\x05\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x00\x00\x02\xff\xff\xff\xff\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00'),type=6) caused by version=0x4,msg_type=0xf,msg_len=0x50,xid=0xc369eeb3,OFPGroupMod(buckets=[OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=5,type=0)],len=32,watch_group=4294967295,watch_port=5,weight=0), OFPBucket(actions=[OFPActionOutput(len=16,max_len=0,port=2,type=0)],len=32,watch_group=4294967295,watch_port=2,weight=0)],command=0,group_id=1002,type=3)


----------------------------------------------------------------------
Ran 2 tests in 80.138s
```

I had to run the tests in fakehw mode (which causes the same switch instance to be used for back to back tests), the regular travis tests don't see the problem as each test is run against a clean switch.

Here is my command to replicate:

```
docker run --rm --privileged --cap-add=NET_ADMIN \
       --sysctl net.ipv6.conf.all.disable_ipv6=0 \
       -v /var/local/lib/docker:/var/lib/docker \
       -v /tmp/faucet-pip-cache:/var/tmp/pip-cache \
       -e HWTESTS=1 \
       -e FAUCET_TESTS="FaucetNailedFailoverForwardingOrderedTest FaucetNailedFailoverForwardingTestcat" \
       -t faucet/tests || exit 1
```

--- a/faucet/valve_of.py
+++ b/faucet/valve_of.py
@@ -983,10 +983,17 @@ def valve_flowreorder(input_ofmsgs, use_barriers=True):
     output_ofmsgs = []
     by_kind = _partition_ofmsgs(input_ofmsgs)
 
-    # Suppress all other deletes if a global delete is present.
+    # Suppress all other relevant deletes if a global delete is present.
     delete_global_ofmsgs = by_kind.get('deleteglobal', [])
     if delete_global_ofmsgs:
-        by_kind['delete'] = []
+        global_types = []
+        for ofmsg in delete_global_ofmsgs:
+            global_types.append(type(ofmsg))
+        new_delete = []
+        for ofmsg in by_kind.get('delete', []):
+            if type(ofmsg) not in global_types:
+                new_delete.append(ofmsg)
+        by_kind['delete'] = new_delete
 
     for kind, random_order, suggest_barrier in _OFMSG_ORDER:
         ofmsgs = dedupe_ofmsgs(by_kind.get(kind, []), random_order)

--- a/tests/unit/faucet/test_valve_stack.py
+++ b/tests/unit/faucet/test_valve_stack.py
@@ -2155,5 +2155,75 @@ dps:
         self.assertTrue(dp_obj.is_stack_edge())
 
 
+class GroupDeleteACLTestCase(ValveTestBases.ValveTestSmall):
+    """Test that a group ACL creates a groupdel for the group_id"""
+
+    CONFIG = """
+acls:
+    group-acl:
+        - rule:
+            dl_dst: "0e:00:00:00:02:02"
+            actions:
+                output:
+                    failover:
+                        group_id: 1001
+                        ports: [2, 3]
+vlans:
+    vlan100:
+        vid: 100
+dps:
+    s1:
+        dp_id: 0x1
+        hardware: 'GenericTFM'
+        interfaces:
+            1:
+                native_vlan: vlan100
+                acls_in: [group-acl]
+            2:
+                native_vlan: vlan100
+            3:
+                native_vlan: vlan100
+"""
+
+    def setUp(self):
+        self.setup_valve(self.CONFIG)
+
+    def check_groupmods_exist(self, ofmsgs, groupdel_exists=True):
+        """Test that the ACL groupmods exist when expected"""
+        groupdel = None
+        groupmod = None
+        for ofmsg in ofmsgs:
+            if valve_of.is_groupdel(ofmsg) and not valve_of.is_global_groupdel(ofmsg):
+                groupdel = ofmsg
+            elif valve_of.is_groupmod(ofmsg):
+                groupmod = ofmsg
+        self.assertIsNotNone(groupmod)
+        if groupdel_exists:
+            self.assertIsNotNone(groupdel)
+            self.assertTrue(groupdel.group_id, 1001)
+        else:
+            self.assertIsNone(groupdel)
+
+    def test_groupdel_exists(self):
+        """Test valve_flowreorder doesn't remove groupmods unless expected"""
+        valve = self.valves_manager.valves[0x1]
+        port = valve.dp.ports[1]
+        ofmsgs = valve.acl_manager.add_port(port)
+        self.check_groupmods_exist(valve_of.valve_flowreorder(ofmsgs))
+        global_flowmod = valve_of.flowmod(
+            0, ofp.OFPFC_DELETE, ofp.OFPTT_ALL,
+            0, ofp.OFPP_CONTROLLER, ofp.OFPP_CONTROLLER,
+            valve_of.match_from_dict({}), [], 0, 0, 0)
+        self.check_groupmods_exist(
+            valve_of.valve_flowreorder(ofmsgs + [global_flowmod]))
+        global_metermod = valve_of.meterdel()
+        self.check_groupmods_exist(
+            valve_of.valve_flowreorder(ofmsgs + [global_flowmod, global_metermod]))
+        global_groupmod = valve_of.groupdel()
+        self.check_groupmods_exist(
+            valve_of.valve_flowreorder(
+                ofmsgs + [global_flowmod, global_metermod, global_groupmod]), False)
+
+
 if __name__ == "__main__":
     unittest.main()  # pytype: disable=module-attr

